database.cc

```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
+                // KS directory
+                sstring ks_name = de.name;
+
+                return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable {
+                    // CF directory
+                    sstring cf_name = de.name.substr(0, de.name.find("-"));
+                    fs::path cf_dir = parent_dir / de.name;
+
+                    return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name, cf_name = std::move(cf_name), cf_dir = std::move(cf_dir)] (fs::path parent_dir, directory_entry de) mutable {
+                        // "snapshots" directory
+
+                        return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &details, ks_name, cf_name, cf_dir] (fs::path parent_dir, directory_entry de) {
+                            sstring snapshot_name = de.name;
+                            std::pair<sstring, sstring> cf_map_key = { ks_name, cf_name };
+                            std::unordered_map<std::pair<sstring, sstring>, snapshot_ctl::snapshot_details, utils::tuple_hash> cf_map = {{ cf_map_key, {0, 0, cf_name, ks_name} }};
+                            details.emplace(snapshot_name, std::move(cf_map));
+
+                            return lister::scan_dir(parent_dir / de.name,  { directory_entry_type::regular }, [this, &details, ks_name, cf_name, cf_map_key, cf_dir, snapshot_name = std::move(snapshot_name)] (fs::path snapshot_dir, directory_entry de) {
+                                return io_check(file_size, (snapshot_dir / de.name).native()).then([this, cf_dir, cf_map_key, &details, snapshot_name, snapshot_dir, name = de.name] (auto size) {
+                                    // The manifest is the only file expected to be in this directory not belonging to the SSTable.
+                                    // For it, we account the total size, but zero it for the true size calculation.
+                                    //
+                                    // All the others should just generate an exception: there is something wrong, so don't blindly
+                                    // add it to the size.
+                                    if (name != "manifest.json" && name != "schema.cql") {
+                                        details.at(snapshot_name).at(cf_map_key).total += size;
+                                    } else {
+                                        size = 0;
+                                    }
+                                    return io_check(file_size, (fs::path(cf_dir) / name).native()).then_wrapped([cf_map_key, &details, snapshot_name, size] (auto fut) {
+                                        try {
+                                            // File exists in the main SSTable directory. Snapshots are not contributing to size
```

Comment:
"File exists in the main SSTable directory"
If I understand correctly, we don't know that yet, we will know it after we leave the `catch` block below. The comment at this place is a little confusing, IMO it should be moved after the `catch`.

Also "Snapshots are not contributing to size" - can you elaborate why? (by continuing this comment)